src/factory-reset/factory-reset-tool.c

```diff
@@ -0,0 +1,374 @@
+                return 0;
+        }
+
+        if (!is_efi_boot()) {
+                if (!arg_quiet)
+                        log_info("Not an EFI boot, cannot remove FactoryResetMode EFI variable, not cancelling.");
+
+                return 0;
+        }
+
+        r = efi_set_variable(EFI_SYSTEMD_VARIABLE_STR("FactoryResetRequest"), /* value= */ NULL, /* size= */ 0);
+        if (r < 0)
+                return log_error_errno(r, "Failed to remove FactoryResetRequest EFI variable: %m");
+
+        if (!arg_quiet)
+                log_info("Factory reset cancelled.");
+
+        return 0;
+}
+
+static int retrigger_block_devices(void) {
+        int r;
+
+        /* Let's retrigger block devices after factory reset is complete: it's quite likely that some
+         * partitions went away or got recreated, and will only be considered relevant once factory reset
+         * mode is left. For example, /dev/disk/gpt-auto-root is like that: it is only created once factory
+         * reset mode is complete. */
+
+        if (!udev_available()) {
+                log_info("Skipping triggering of block devices, as udev is not available.");
+                return 0;
+        }
+
+        _cleanup_(sd_device_enumerator_unrefp) sd_device_enumerator *e = NULL;
+        r = sd_device_enumerator_new(&e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate device enumerator: %m");
+
+        r = sd_device_enumerator_allow_uninitialized(e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable enumeration of uninitialized devices: %m");
+
+        r = sd_device_enumerator_add_match_subsystem(e, "block", /* match = */ true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to filter device enumeration by 'block' subsystem: %m");
+
+        if (!arg_quiet)
+                log_info("Retriggering block devices.");
+
+        FOREACH_DEVICE(e, d) {
+                r = sd_device_trigger(d, SD_DEVICE_CHANGE);
+                if (r < 0)
+                        log_device_debug_errno(d, r, "Failed to trigger block device, ignoring: %m");
+        }
+
+        return 0;
+}
+
+static int verb_complete(int argc, char *argv[], void *userdata) {
+        int r;
+
+        FactoryResetMode f = factory_reset_mode();
+        if (f < 0)
+                return log_error_errno(f, "Failed to dermine factory reset state: %m");
+        log_debug("Current factory state is: %s", factory_reset_mode_to_string(f));
+        if (f != FACTORY_RESET_ON) {
+                log_info("Attempted to leave factory reset mode, even though we are not in factory reset mode. Ignoring.");
+                return 0;
+        }
+
+        r = efi_set_variable(EFI_SYSTEMD_VARIABLE_STR("FactoryResetRequest"), /* value= */ NULL, /* size= */ 0);
+        if (r < 0)
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_WARNING, r,
```

Comment:
Maybe also this one?

Reply:
hmm, this is a real problem, i don't think we should hide that...

i'll condition it with `is_efi_boot()` though, to ensure we don't generate a warning on non-efi systems for this.

Reply:
OK.